stdlib/regex.gr

```diff
@@ -0,0 +1,2983 @@
+    }
+  }
+  let (minLen, maxLen, maxLookbehind) = loop(re)
+  Map.forEach((k, _) => {
+    match(Map.get(k, groupSizes)) {
+      None => void,
+      Some(sz) => {
+        if (sz <= 0) {
+          thrownError = Some(MightBeEmpty)
+        }
+      }
+    }
+  }, mustSizes)
+  match(thrownError) {
+    Some(MightBeEmpty) => Err("`*`, `+`, or `{...}` operand could be empty"),
+    Some(DoesNotMatchBounded) => Err("lookbehind pattern does not match a bounded length"),
+    Some(BackreferenceTooBig) => Err("backreference number is larger than the highest-numbered cluster"),
+    Some(InternalError(re)) => Err("regex validate: Internal error: " ++ toString(re)),
+    None => Ok(maxLookbehind)
+  }
+}
+
+
+/*
+
+=========================
+REGEX MATCHER COMPILATION
+=========================
+
+*/
+
+
+record MatchBuf {
+  matchInput: String,
+  matchInputExploded: Array<Char>,
+}
+
+let makeMatchBuffer = (s) => {
+  {
+    matchInput: s,
+    matchInputExploded: String.explode(s),
+  }
+}
+
+let matchBufMore = (buf: MatchBuf, pos: Number) => {
+  pos < Array.length(buf.matchInputExploded)
+}
+
+let matchBufChar = (buf: MatchBuf, pos: Number) => {
+  if (pos >= Array.length(buf.matchInputExploded)) {
+    Err("end of match buffer reached")
+  } else {
+    Ok(buf.matchInputExploded[pos])
+  }
+}
+
+enum StackElt {
+  SEPositionProducer(Number -> Option<Number>),
+  SESavedGroup(Number, Option<(Number, Number)>),
+}
+
+let done_m = (buf: MatchBuf, pos: Number, start: Number, limit: Number, end: Number, state, stack) => Some(pos)
+let continue_m = (buf: MatchBuf, pos: Number, start: Number, limit: Number, end: Number, state, stack) => {
+  match(stack) {
+    [SEPositionProducer(hd), ..._] => hd(pos),
+    _ => fail "Impossible: continue_m",
+  }
+}
+let limit_m = (buf: MatchBuf, pos: Number, start: Number, limit: Number, end: Number, state, stack) => if (pos == limit) Some(pos) else None
+
+
+let iterateMatcher = (m, size, max) => (buf: MatchBuf, pos: Number, start: Number, limit: Number, end: Number, state, stack) => {
+  let limit = match(max) {
+    Some(max) => min(limit, pos + (max * size)),
+    None => limit,
+  }
+  let rec loop = (pos2, n) => {
+    let pos3 = pos2 + size
+    if (pos3 > limit || !m(buf, pos2, start, limit, end, state, stack)) {
+      (pos2, n, size)
+    } else {
+      loop(pos3, n + 1)
+    }
+  }
+  loop(pos, 0)
+}
+
+// single-char matching
+
+let charMatcher = (toMatch, next_m) => (buf: MatchBuf, pos: Number, start: Number, limit: Number, end: Number, state, stack) => {
+  if ({
+    pos < limit && match(matchBufChar(buf, pos)) {
+      Err(_) => false,
+      Ok(c) => toMatch == c
+    }
+  }) next_m(buf, pos + 1, start, limit, end, state, stack) else None
+}
+
+let charTailMatcher = (toMatch) => (buf: MatchBuf, pos: Number, start: Number, limit: Number, end: Number, state, stack) => {
+  if ({
+    pos < limit && match(matchBufChar(buf, pos)) {
+      Err(_) => false,
+      Ok(c) => toMatch == c
+    }
```

Comment:
To avoid a proliferation of helper functions (the only clean way to do it that I can see will require allocating a bunch of lambdas, and this code path is hot enough that I think the speedup of `match` warrants the sacrificed cleanliness), I am going to resolve this as-is.